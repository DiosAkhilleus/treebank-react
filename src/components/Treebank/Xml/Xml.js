import React from 'react';

import TreebankContext from '../treebank-context';

import Xml from '../../Xml';

const XmlWithContext = () => (
  <TreebankContext.Consumer>
    {({ sentence, active, toggleActive }) => (
      <Xml sentence={sentence} active={active} toggleActive={toggleActive} />
    )}
  </TreebankContext.Consumer>
);

export default XmlWithContext;
