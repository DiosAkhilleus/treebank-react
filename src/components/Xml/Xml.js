import React, { Fragment } from 'react';

import { sentenceType } from '../../types';

import styles from './Xml.module.scss';

const attributesToJsx = (attributes) => (
  Object.entries(attributes).map(([key, value]) => (
    <Fragment key={key}>
      {' '}
      <span className={styles.attribute}>
        {key}
        =
      </span>
      <span className={styles.attributeText}>
        &quot;
        {value}
        &quot;
      </span>
    </Fragment>
  ))
);

const openingTag = (name, attributes, key, close = false, prefix = '') => (
  <Fragment key={key}>
    <span className={styles.bracket}>
      {prefix}
      &lt;
    </span>
    <span className={styles.element}>{name}</span>
    {attributesToJsx(attributes)}
    {close ? ' /' : ''}
    <span className={styles.bracket}>&gt;</span>
    <br />
  </Fragment>
);

const closingTag = (name, key) => (
  <Fragment key={key}>
    <span className={styles.bracket}>&lt;</span>
    <span className={styles.element}>
      /
      {name}
    </span>
    <span className={styles.bracket}>&gt;</span>
  </Fragment>
);

const Xml = ({ sentence }) => (
  <>
    {openingTag('sentence', sentence.$, 'sentence')}
    {sentence.word.map((word) => (
      openingTag('word', word.$, word.$.id, true, <>&nbsp;&nbsp;&nbsp;&nbsp;</>)
    ))}
    {closingTag('sentence', 'sentence-close')}
  </>
);

Xml.propTypes = {
  sentence: sentenceType.isRequired,
};

export default Xml;
