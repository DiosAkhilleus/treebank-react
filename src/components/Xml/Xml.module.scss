.xml {
  background-color: rgb(242, 242, 242);

  padding-top: 0.2rem;
  padding-bottom: 0.2rem;
  padding-left: 0.4rem;
  padding-right: 0.4rem;
}

.word {
  cursor: pointer;

  &:hover {
    background-color: #5BC8DC;
  }
}

.active {
  background-color: rgb(246, 217, 24);
}

.bracket {
  color: rgb(0, 0, 255);
}

.element {
  color: rgb(128, 0, 128);
}

.attribute {
  color: rgb(0, 0, 0);
}

.attributeText {
  color: rgb(255, 0, 0);
}
