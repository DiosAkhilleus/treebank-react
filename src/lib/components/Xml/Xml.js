import React, { Fragment } from 'react';
import { func, instanceOf } from 'prop-types';

import { sentenceType, wordType } from '../../types';

import styles from './Xml.module.scss';

const attributesToJsx = (attributes) => (
  Object.entries(attributes).map(([key, value]) => (
    <Fragment key={key}>
      {' '}
      <span className={styles.attribute}>
        {key}
        =
      </span>
      <span className={styles.attributeText}>
        &quot;
        {value}
        &quot;
      </span>
    </Fragment>
  ))
);

const openingTag = (name, attributes, key, close = false, prefix = '') => (
  <Fragment key={key}>
    <span className={styles.bracket}>
      {prefix}
      &lt;
    </span>
    <span className={styles.element}>{name}</span>
    {attributesToJsx(attributes)}
    {close ? ' /' : ''}
    <span className={styles.bracket}>&gt;</span>
    <br />
  </Fragment>
);

const closingTag = (name, key) => (
  <Fragment key={key}>
    <span className={styles.bracket}>&lt;</span>
    <span className={styles.element}>
      /
      {name}
    </span>
    <span className={styles.bracket}>&gt;</span>
  </Fragment>
);

const renderWord = (word, active, toggleActive, highlight) => {
  const { $: { id } } = word;
  const classes = [styles.word];
  const onClick = () => {
    toggleActive(id);
  };
  const onKeyDown = (event) => {
    const { key } = event;

    if (key === 'Enter') {
      onClick(event);
    }
  };

  if (active && active.$.id === id) {
    classes.push(styles.active);
  } else if (highlight.has(id)) {
    classes.push(styles.highlight);
  }

  return (
    <div
      key={id}
      role="button"
      tabIndex="0"
      className={classes.join(' ')}
      onClick={onClick}
      onKeyDown={onKeyDown}
    >
      {openingTag('word', word.$, id, true, <>&nbsp;&nbsp;&nbsp;&nbsp;</>)}
    </div>
  );
};

const Xml = ({
  sentence, active, toggleActive, highlight,
}) => (
  <div className={styles.xml}>
    {openingTag('sentence', sentence.$, 'sentence')}
    {sentence.word.map((word) => (
      renderWord(word, active, toggleActive, highlight)
    ))}
    {closingTag('sentence', 'sentence-close')}
  </div>
);

Xml.propTypes = {
  sentence: sentenceType.isRequired,
  active: wordType,
  toggleActive: func,
  highlight: instanceOf(Set),
};

Xml.defaultProps = {
  active: null,
  toggleActive: () => {},
  highlight: new Set(),
};

export default Xml;
