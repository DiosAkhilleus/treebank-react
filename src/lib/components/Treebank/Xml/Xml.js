import React from 'react';

import SentenceContext from '../sentence-context';

import Xml from '../../Xml';

const XmlWithContext = () => (
  <SentenceContext.Consumer>
    {({
      sentence, active, toggleActive, highlight,
    }) => (
      <Xml
        sentence={sentence}
        active={active}
        toggleActive={toggleActive}
        highlight={highlight}
      />
    )}
  </SentenceContext.Consumer>
);

export default XmlWithContext;
