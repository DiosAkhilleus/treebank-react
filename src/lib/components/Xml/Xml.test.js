import React from 'react';
import renderer from 'react-test-renderer';

import Xml from './Xml';

it('renders a sentence as XML', () => {
  const sentence = {
    $: {
      id: '1', document_id: 'test', subdoc: '2', span: '',
    },
    word: [
      {
        $: {
          id: '1', form: 'οὐκ', postag: 'd--------', relation: 'AuxZ', head: '1',
        },
      },
      {
        $: {
          id: '2', form: 'ᾔδη', lemma: 'οἶδα', postag: 'v1slia---', relation: 'PRED', head: '0',
        },
      },
    ],
  };

  const component = (
    <Xml sentence={sentence} />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});

it('displays highlighted and active words differently', () => {
  const sentence = {
    $: {
      id: '1', document_id: 'test', subdoc: '2', span: '',
    },
    word: [
      { $: { id: '1', form: 'Ἡροδότου', postag: 'n-s---mg-' } },
      { $: { id: '2', form: 'Ἁλικαρνησσέος', postag: 'n-s---mg-' } },
      { $: { id: '3', form: 'ἱστορίης', postag: 'n-s---fg-' } },
      { $: { id: '4', form: 'ἀπόδεξις', postag: 'n-s---fn-' } },
    ],
  };

  const component = (
    <Xml
      sentence={sentence}
      active={{ $: { id: '1' } }}
      highlight={new Set(['1', '3'])}
    />
  );
  const tree = renderer.create(component).toJSON();

  expect(tree).toMatchSnapshot();
});
