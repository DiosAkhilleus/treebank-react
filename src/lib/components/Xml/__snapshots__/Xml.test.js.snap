// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`displays highlighted and active words differently 1`] = `
<div
  className="xml"
>
  <span
    className="bracket"
  >
    
    &lt;
  </span>
  <span
    className="element"
  >
    sentence
  </span>
   
  <span
    className="attribute"
  >
    id
    =
  </span>
  <span
    className="attributeText"
  >
    "
    1
    "
  </span>
   
  <span
    className="attribute"
  >
    document_id
    =
  </span>
  <span
    className="attributeText"
  >
    "
    test
    "
  </span>
   
  <span
    className="attribute"
  >
    subdoc
    =
  </span>
  <span
    className="attributeText"
  >
    "
    2
    "
  </span>
   
  <span
    className="attribute"
  >
    span
    =
  </span>
  <span
    className="attributeText"
  >
    "
    
    "
  </span>
  
  <span
    className="bracket"
  >
    &gt;
  </span>
  <br />
  <div
    className="word active"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <span
      className="bracket"
    >
          
      &lt;
    </span>
    <span
      className="element"
    >
      word
    </span>
     
    <span
      className="attribute"
    >
      id
      =
    </span>
    <span
      className="attributeText"
    >
      "
      1
      "
    </span>
     
    <span
      className="attribute"
    >
      form
      =
    </span>
    <span
      className="attributeText"
    >
      "
      Ἡροδότου
      "
    </span>
     
    <span
      className="attribute"
    >
      postag
      =
    </span>
    <span
      className="attributeText"
    >
      "
      n-s---mg-
      "
    </span>
     /
    <span
      className="bracket"
    >
      &gt;
    </span>
    <br />
  </div>
  <div
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <span
      className="bracket"
    >
          
      &lt;
    </span>
    <span
      className="element"
    >
      word
    </span>
     
    <span
      className="attribute"
    >
      id
      =
    </span>
    <span
      className="attributeText"
    >
      "
      2
      "
    </span>
     
    <span
      className="attribute"
    >
      form
      =
    </span>
    <span
      className="attributeText"
    >
      "
      Ἁλικαρνησσέος
      "
    </span>
     
    <span
      className="attribute"
    >
      postag
      =
    </span>
    <span
      className="attributeText"
    >
      "
      n-s---mg-
      "
    </span>
     /
    <span
      className="bracket"
    >
      &gt;
    </span>
    <br />
  </div>
  <div
    className="word highlight"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <span
      className="bracket"
    >
          
      &lt;
    </span>
    <span
      className="element"
    >
      word
    </span>
     
    <span
      className="attribute"
    >
      id
      =
    </span>
    <span
      className="attributeText"
    >
      "
      3
      "
    </span>
     
    <span
      className="attribute"
    >
      form
      =
    </span>
    <span
      className="attributeText"
    >
      "
      ἱστορίης
      "
    </span>
     
    <span
      className="attribute"
    >
      postag
      =
    </span>
    <span
      className="attributeText"
    >
      "
      n-s---fg-
      "
    </span>
     /
    <span
      className="bracket"
    >
      &gt;
    </span>
    <br />
  </div>
  <div
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <span
      className="bracket"
    >
          
      &lt;
    </span>
    <span
      className="element"
    >
      word
    </span>
     
    <span
      className="attribute"
    >
      id
      =
    </span>
    <span
      className="attributeText"
    >
      "
      4
      "
    </span>
     
    <span
      className="attribute"
    >
      form
      =
    </span>
    <span
      className="attributeText"
    >
      "
      ἀπόδεξις
      "
    </span>
     
    <span
      className="attribute"
    >
      postag
      =
    </span>
    <span
      className="attributeText"
    >
      "
      n-s---fn-
      "
    </span>
     /
    <span
      className="bracket"
    >
      &gt;
    </span>
    <br />
  </div>
  <span
    className="bracket"
  >
    &lt;
  </span>
  <span
    className="element"
  >
    /
    sentence
  </span>
  <span
    className="bracket"
  >
    &gt;
  </span>
</div>
`;

exports[`renders a sentence as XML 1`] = `
<div
  className="xml"
>
  <span
    className="bracket"
  >
    
    &lt;
  </span>
  <span
    className="element"
  >
    sentence
  </span>
   
  <span
    className="attribute"
  >
    id
    =
  </span>
  <span
    className="attributeText"
  >
    "
    1
    "
  </span>
   
  <span
    className="attribute"
  >
    document_id
    =
  </span>
  <span
    className="attributeText"
  >
    "
    test
    "
  </span>
   
  <span
    className="attribute"
  >
    subdoc
    =
  </span>
  <span
    className="attributeText"
  >
    "
    2
    "
  </span>
   
  <span
    className="attribute"
  >
    span
    =
  </span>
  <span
    className="attributeText"
  >
    "
    
    "
  </span>
  
  <span
    className="bracket"
  >
    &gt;
  </span>
  <br />
  <div
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <span
      className="bracket"
    >
          
      &lt;
    </span>
    <span
      className="element"
    >
      word
    </span>
     
    <span
      className="attribute"
    >
      id
      =
    </span>
    <span
      className="attributeText"
    >
      "
      1
      "
    </span>
     
    <span
      className="attribute"
    >
      form
      =
    </span>
    <span
      className="attributeText"
    >
      "
      οὐκ
      "
    </span>
     
    <span
      className="attribute"
    >
      postag
      =
    </span>
    <span
      className="attributeText"
    >
      "
      d--------
      "
    </span>
     
    <span
      className="attribute"
    >
      relation
      =
    </span>
    <span
      className="attributeText"
    >
      "
      AuxZ
      "
    </span>
     
    <span
      className="attribute"
    >
      head
      =
    </span>
    <span
      className="attributeText"
    >
      "
      1
      "
    </span>
     /
    <span
      className="bracket"
    >
      &gt;
    </span>
    <br />
  </div>
  <div
    className="word"
    onClick={[Function]}
    onKeyDown={[Function]}
    role="button"
    tabIndex="0"
  >
    <span
      className="bracket"
    >
          
      &lt;
    </span>
    <span
      className="element"
    >
      word
    </span>
     
    <span
      className="attribute"
    >
      id
      =
    </span>
    <span
      className="attributeText"
    >
      "
      2
      "
    </span>
     
    <span
      className="attribute"
    >
      form
      =
    </span>
    <span
      className="attributeText"
    >
      "
      ᾔδη
      "
    </span>
     
    <span
      className="attribute"
    >
      lemma
      =
    </span>
    <span
      className="attributeText"
    >
      "
      οἶδα
      "
    </span>
     
    <span
      className="attribute"
    >
      postag
      =
    </span>
    <span
      className="attributeText"
    >
      "
      v1slia---
      "
    </span>
     
    <span
      className="attribute"
    >
      relation
      =
    </span>
    <span
      className="attributeText"
    >
      "
      PRED
      "
    </span>
     
    <span
      className="attribute"
    >
      head
      =
    </span>
    <span
      className="attributeText"
    >
      "
      0
      "
    </span>
     /
    <span
      className="bracket"
    >
      &gt;
    </span>
    <br />
  </div>
  <span
    className="bracket"
  >
    &lt;
  </span>
  <span
    className="element"
  >
    /
    sentence
  </span>
  <span
    className="bracket"
  >
    &gt;
  </span>
</div>
`;
